### ISSUE TITLE
Feature request: implement some hashing functions in RainerScript

### ISSUE THREAD

**Issue by jay7x** [rsyslog/rsyslog#2273]
Can we have some hashing functions implemented? It's frequently used to distribute flow evenly between multiple rulesets/receivers. For now it's hard to do using RainerScript only w/o external modules.

I have no big experience with hashing functions so I can't propose final list of good candidates. But I'd say it's good to have at least one CRC/Checksum function and one cryptographic function.

Example of usage:
```
set $.hash_partition = hash_func($fromhost);
set $.partition = exec_template("get_last_char_of_hash_result");
call_indirect "partition_" & $.partition;
```

May be it's good idea to have some kind of `substr(string)` function as well to avoid using templates for string manipulation. Or introduce some function to apply property replacer expressions as function argument (w/o need to define template). But this is another feature request :)

**Comment by deoren** [comment]
> @jay7x: May be it's good idea to have some kind of substr(string) function as well to avoid using templates for string manipulation.

Good news!

http://www.rsyslog.com/doc/master/rainerscript/functions.html#substring-str-start-substringlength

This should be available in the 8.32.0 stable release (#2191).

**Comment by jay7x** [comment]
WOW! I've missed this somehow! Really great news! 👍

**Comment by rgerhards** [comment]
@jay7x besides the good news, I would strongly recommend to open one issue per function request. If they are small, people (including me) tend to pick up issues quicker. so up-to-the-point issue are most often implemented quicker. Just my 2cts...

**Comment by rgerhards** [comment]
@jay7x ... I think the generic "hashing functions" is fine, just adding a second one like substr I meant.

**Comment by rgerhards** [comment]
looks like https://github.com/rsyslog/rsyslog/pull/2465 implements this

**Comment by rgerhards** [comment]
closed via #2465
